File: tsp.cpp
#include <algorithm>
#include <iostream>
#include <limits>
#include <numeric>
#include <vector>
using namespace std;
// Define the objective function for TSP using a distance matrix
double objective_function(const vector<int>& solution, const vector<vector<double>>& distance_matrix) {
double total_distance = 0.0;
for (size_t i = 0; i < solution.size() - 1; ++i) {
total_distance += distance_matrix[solution[i]][solution[i + 1]];
}
// Add distance from the last city back to the starting city
total_distance += distance_matrix[solution.back()][solution[0]];
return total_distance;
}
// Modify the get_neighbors function to generate TSP-specific neighbors
vector<vector<int>> get_neighbors(const vector<int>& solution) {
vector<vector<int>> neighbors;
for (size_t i = 0; i < solution.size() - 1; ++i) {
for (size_t j = i + 1; j < solution.size(); ++j) {
vector<int> neighbor = solution;
swap(neighbor[i], neighbor[j]);
neighbors.push_back(neighbor);
}
}
return neighbors;
}
// Rest of the code remains the same
int main() {
// Example usage
// Define the distance matrix
vector<vector<double>> distance_matrix = {
{0.0, 2.0, 3.0, 4.0, 5.0},
{2.0, 0.0, 1.0, 2.0, 3.0},
{3.0, 1.0, 0.0, 1.0, 2.0},
{4.0, 2.0, 1.0, 0.0, 1.0},
{5.0, 3.0, 2.0, 1.0, 0.0}
};
// Provide an initial solution
vector<int> initial_solution = { 0, 1, 2, 3, 4 }; // Assuming starting from city 0
int max_iterations = 100;
int tabu_list_size = 10;
vector<int> best_solution = tabu_search(initial_solution, max_iterations, tabu_list_size);
cout << "Best solution:";
for (int val : best_solution) {
cout << " " << val;
}
cout << endl;
cout << "Best solution total distance: " << objective_function(best_solution, distance_matrix) << endl;
return 0;
}

File: tsp.js
function objectiveFunction(solution) {
    // TODO: Implement your objective function here
    // The objective function should evaluate
    // the quality of a given solution and
    // return a numerical value representing
    // the solution's fitness
    // Example: return solution.reduce((sum, val) => sum + val, 0);
    return solution.reduce((sum, val) => sum + val, 0);
}

function getNeighbors(solution) {
    const neighbors = [];
    for (let i = 0; i < solution.length; i++) {
        for (let j = i + 1; j < solution.length; j++) {
            const neighbor = [...solution];
            // Swap elements at indices i and j
            [neighbor[i], neighbor[j]] = [neighbor[j], neighbor[i]];
            neighbors.push(neighbor);
        }
    }
    return neighbors;
}

function tabuSearch(initialSolution, maxIterations, tabuListSize) {
    let bestSolution = [...initialSolution];
    let currentSolution = [...initialSolution];
    const tabuList = [];

    for (let iter = 0; iter < maxIterations; iter++) {
        const neighbors = getNeighbors(currentSolution);
        let bestNeighbor = [];
        let bestNeighborFitness = Number.MAX_VALUE;

        for (const neighbor of neighbors) {
            if (!tabuList.some(entry => JSON.stringify(entry) === JSON.stringify(neighbor))) {
                const neighborFitness = objectiveFunction(neighbor);
                if (neighborFitness < bestNeighborFitness) {
                    bestNeighbor = [...neighbor];
                    bestNeighborFitness = neighborFitness;
                }
            }
        }

        if (bestNeighbor.length === 0) {
            // No non-tabu neighbors found, terminate the search
            break;
        }

        currentSolution = [...bestNeighbor];
        tabuList.push([...bestNeighbor]);

        if (tabuList.length > tabuListSize) {
            // Remove the oldest entry from the tabu list if it exceeds the size
            tabuList.shift();
        }

        if (objectiveFunction(bestNeighbor) < objectiveFunction(bestSolution)) {
            // Update the best solution if the current neighbor is better
            bestSolution = [...bestNeighbor];
        }
    }

    return bestSolution;
}

// Example usage
const initialSolution = [1, 2, 3, 4, 5];
const maxIterations = 100;
const tabuListSize = 10;

const bestSolution = tabuSearch(initialSolution, maxIterations, tabuListSize);
console.log("Best solution:", bestSolution.join(" "));
console.log("Best solution fitness:", objectiveFunction(bestSolution));
